### PR TITLE
io_buffer.c: Allow copies between overlapping buffers with #copy and #set_string

### PR DESCRIPTION
**Pull request by hanazuki**
https://bugs.ruby-lang.org/issues/20745

### CODE REVIEW

**Comment by casperisfine** [comment]
Please include `[Bug #20745]` in your commit message so that if this is merged it's properly associated to the issue on Redmine, that makes it much easier for backport and such.

File: test/ruby/test_io_buffer.rb [hunk 1]
```diff
@@ -572,4 +572,54 @@ def test_private
...
+    buf = IO::Buffer.new(0)
+    assert_equal true, buf.null?
+    buf.copy(IO::Buffer.for('a'), 0, 0)
+    assert_equal true, buf.empty?
+  end
+
+  def test_copy_null_source
+    buf = IO::Buffer.for('a').dup
+    src = IO::Buffer.new(0)
+    assert_equal true, src.null?
```
[the remark this answers is not in the document]
Comment (hanazuki) [reply test/ruby/test_io_buffer.rb]: Didnt know. Thanks.

**Comment by hanazuki** [comment]
I've revised my patch. The CI failure (io-wait one) seems unrelated to this change.

File: io_buffer.c [hunk 2]
```diff
@@ -2471,6 +2473,19 @@ rb_io_buffer_initialize_copy(VALUE self, VALUE source)
  *    buffer = IO::Buffer.new(2)
  *    buffer.copy(IO::Buffer.for('test'), 0)
  *    # in `copy': Specified offset+length is bigger than the buffer size! (ArgumentError)
+ *
+ *  It is safe to copy between the memory regions that overlaps each other.
+ *  In such a case, the data is copied as if the data is first copied from the source buffer to
```
Comment (byroot) [inline io_buffer.c:2478]:
```suggestion
 *  It is safe to copy between memory regions that overlaps each other.
 *  In such case, the data is copied as if the data was first copied from the source buffer to
```

**Review by byroot — APPROVED** [review]
I'm no expert in `memove` semantic, but based on the manpage and the implementations you linked to, it seems like the sensible thing to do.

Just some minor suggestions on the documentation.

**Comment by hanazuki** [comment]
Thank you @byroot for your review. I've incorporated your suggestion on the grammar. Can you merge this patch? cc @nobu

**Comment by byroot** [comment]
I believe @ioquatix maintains this class, so I added him as reviewer so he has a chance to chime in if necessary, but if not I'll merge tomorrow.

**Comment by ioquatix** [comment]
Thanks I will review it today.

**Review by ioquatix — APPROVED** [review]
This looks great to me. Thanks for your attention to writing good tests. I've rebased and enabled auto-merge.

**Comment by ioquatix** [comment]
If you are interested in another interesting work in the same area of code, I wonder if we can find out at what threshold (length) we should release the GVL. I think we can assume most modern CPUs are between 50-100GB/s memory bandwidth. How that translates to mem(cpy/move) performance I'm less sure, but I think we should aim to release the GVL if an operation is going to take more than some unit of time, e.g. 1ms.

**Comment by hanazuki** [comment]
Thank you, Samuel. Releasing GVL for large memcopies sounds like an interesting idea. However, I don't have enough insights for that, as my current use case of IO::Buffer is mainly for small network packets, where most data can be handled in a zero-copy manner and the constant overheads of tiny operations are dominant.

**Comment by ioquatix** [comment]
Is your use of `IO::Buffer` public code? I would like to take a look at how you are using it.

**Comment by ioquatix** [comment]
Here is an implementation of the nogvl buffer memmove: https://github.com/ruby/ruby/pull/12021

**Comment by ioquatix** [comment]
@casperisfine 

> Please include [Bug #20745] in your commit message so that if this is merged it's properly associated to the issue on Redmine, that makes it much easier for backport and such.

I didn't know this was the magic cheat code to unlock proper integration!!! Thanks for mentioning it :p

**Comment by hanazuki** [comment]
> Is your use of IO::Buffer public code?

Yes, one is available at <https://github.com/sorah/xlat> (a joint project by sorah and me). This is an IPv6/IPv4 translator, so we do many small modifications in the headers, move around the packet payload without copy, and write them together with [writev](https://github.com/sorah/xlat/blob/main/ext/xlat/io_buffer_ext/src/lib.rs).

**Comment by ioquatix** [comment]
At one point I wanted to remove writev from Ruby as libc does a poor job of implementing it. However `io_uring` does appear to have a direct code path for it, so I was thinking of introducing it again. Maybe we can add support for it directly, e.g. `io.writev(array_of_io_buffer)` or `IO::Buffer.writev(io, *buffers)`. I don't have any immediate plan for it, but it's great to see your use case.

**Comment by ioquatix** [comment]
Also, I guess you tested performance between `String` and `IO::Buffer`. Did you get any good results one way or another?

**Comment by hanazuki** [comment]
We need writev for Xlat not only for performance but also for dealing with IO with datagram semantics (TUN device), where we have to write a whole IP datagram at once, and so writev(2) cannot be replaced with multiple calls of write(2).

>  Also, I guess you tested performance between String and IO::Buffer. Did you get any good results one way or another?

It was something like 250 Mbps (String) vs. 340 Mbps (IO::Buffer) with some generated TCP traffic. I guess the improvement is mostly due to the reduction of copies of packet payload (~ 1400 bytes per packet).

> Maybe we can add support for it directly, e.g. io.write(array_of_io_buffer) or IO::Buffer.writev(io, *buffers)

It would be convenient if IO methods could take IO::Buffer directly instead of Strings. But, I suspect that even `IO::Buffer#slice` is slow for my case, as it needs allocation in the heap, and am considering trying something like `writev(io, [buf1, off1, len1], [buf2, off2, len2], ...)`. Completely not sure if it's worth the effort, though.

**Comment by ioquatix** [comment]
That's very reasonable.

In other words, `slice` might be too much of an overhead for `iov` construction (even if it's zero copy, it still allocates a Ruby object).

In other words, we might need an interface like `IO::BufferIOV` (bad name, but I'm sure you get the idea).

**Comment by hanazuki** [comment]
That interface of IO::Vector looks good. I will experiment and let you know if it works well.

**Comment by ioquatix** [comment]
(such a design is compatible with the fiber scheduler so I'm also okay with that aspect).

**Pull request merged**